addons/sale_expense/models/hr_expense_sheet.py

```diff
@@ -15,38 +16,90 @@ def _compute_sale_order_count(self):
         if not expensed_amls:
             return self.env['sale.order.line']
+
+        # Get the sale orders linked to the related expenses
         aml_to_so_map = expensed_amls._sale_determine_order()
         sale_order_ids = tuple(set(aml_to_so_map[aml.id].id for aml in expensed_amls))
-        aml_sol_unit_price_map = dict(expensed_amls.mapped(lambda aml: (aml.id, aml._sale_get_invoice_price(aml_to_so_map[aml.id]))))
+
+        # Map the expense aml to the unit price a related sale order line would have, as it differs from the expense aml's price
+        aml_sol_unit_price_map = dict(expensed_amls.mapped(
+            lambda aml: (aml.id, aml.currency_id.round(aml._sale_get_invoice_price(aml_to_so_map[aml.id])))
+        ))
+
+        # Get the product ids, quantities and names of the expensed account move lines. Used to filter the sale order lines
         product_ids = tuple(expensed_amls.product_id.ids)
         quantities = tuple(expensed_amls.mapped('quantity'))
         names = tuple(expensed_amls.mapped('name'))
+
         self.env['sale.order.line'].flush_model(['order_id', 'product_id', 'product_uom_qty', 'price_unit', 'name'])
+        self.env['res.company'].flush_model(['currency_id'])
+        self.env['res.currency'].flush_model(['rounding'])
         query = """
-            SELECT 
-                DISTINCT ON (sol.order_id, sol.product_id, sol.product_uom_qty, sol.price_unit, sol.name)
-                sol.order_id, sol.product_id, sol.product_uom_qty, sol.price_unit, sol.name, sol.id
+            SELECT
+                sol.order_id,
+                sol.product_id,
+                sol.product_uom_qty,
+                sol.price_unit,
+                sol.name,
+                sol.id,
+                CASE 
+                    WHEN sol.currency_id IS NOT NULL THEN 
+                        currency.rounding
+                    ELSE
+                        company_currency.rounding
+                END AS rounding
             FROM sale_order_line AS sol
+            JOIN res_company AS company ON sol.company_id = company.id
+            JOIN res_currency AS company_currency ON company.currency_id = company_currency.id
+            LEFT JOIN res_currency AS currency ON sol.currency_id = currency.id
             WHERE sol.is_expense = TRUE
                 AND sol.order_id IN %s
                 AND sol.product_id IN %s
                 AND sol.product_uom_qty IN %s
-                AND sol.price_unit IN %s
                 AND sol.name IN %s
-            ORDER BY sol.order_id, sol.product_id, sol.product_uom_qty, sol.price_unit, sol.name
+            GROUP BY (
+                company_currency.rounding,
+                currency.rounding,
+                sol.order_id,
+                sol.product_id,
+                sol.price_unit,
+                sol.product_uom_qty,
+                sol.name,
```

Comment:
```suggestion
```

`sol.id` makes all of these implicit thanks to the primary key/unique constraint